include/deal.II/lac/precondition.h

```diff
@@ -1738,61 +1796,82 @@ namespace internal
       const VectorUpdater<Number> &updater;
     };
 
-    // selection for deal.II vector
+    // selection for diagonal matrix around deal.II vector
     template <typename Number>
     inline
     void
     vector_updates (const ::dealii::Vector<Number> &src,
-                    const ::dealii::Vector<Number> &matrix_diagonal_inverse,
+                    const DiagonalMatrix<::dealii::Vector<Number> > &jacobi,
                     const bool    start_zero,
                     const double  factor1,
                     const double  factor2,
                     ::dealii::Vector<Number> &update1,
                     ::dealii::Vector<Number> &update2,
+                    ::dealii::Vector<Number> &,
                     ::dealii::Vector<Number> &dst)
     {
-      VectorUpdater<Number> upd(src.begin(), matrix_diagonal_inverse.begin(),
+      VectorUpdater<Number> upd(src.begin(), jacobi.get_vector().begin(),
                                 start_zero, factor1, factor2,
                                 update1.begin(), update2.begin(), dst.begin());
       VectorUpdatesRange<Number>(upd, src.size());
     }
 
-    // selection for parallel deal.II vector
+    // selection for diagonal matrix around parallel deal.II vector
     template <typename Number>
     inline
     void
     vector_updates (const LinearAlgebra::distributed::Vector<Number> &src,
-                    const LinearAlgebra::distributed::Vector<Number> &matrix_diagonal_inverse,
+                    const DiagonalMatrix<LinearAlgebra::distributed::Vector<Number> > &jacobi,
                     const bool    start_zero,
                     const double  factor1,
                     const double  factor2,
                     LinearAlgebra::distributed::Vector<Number> &update1,
                     LinearAlgebra::distributed::Vector<Number> &update2,
+                    LinearAlgebra::distributed::Vector<Number> &,
                     LinearAlgebra::distributed::Vector<Number> &dst)
     {
-      VectorUpdater<Number> upd(src.begin(), matrix_diagonal_inverse.begin(),
+      VectorUpdater<Number> upd(src.begin(), jacobi.get_vector().begin(),
                                 start_zero, factor1, factor2,
                                 update1.begin(), update2.begin(), dst.begin());
       VectorUpdatesRange<Number>(upd, src.local_size());
     }
 
-    template <typename VectorType>
-    struct DiagonalPreconditioner
+    template <typename MatrixType, typename VectorType, typename PreconditionerType>
+    inline
+    void
+    initialize_preconditioner(const MatrixType   &matrix,
+                              PreconditionerType &preconditioner,
+                              VectorType         &diagonal_inverse)
     {
-      DiagonalPreconditioner (const VectorType &vector)
-        :
-        diagonal_vector(vector)
-      {}
+      AssertDimension(preconditioner.m(), matrix.m());
```

Comment:
This function should do nothing because it is the case where we have a `PreconditionerType` object that is not just a diagonal matrix derived from a vector, so the user must have provided a suitable operation. The other `initialize_preconditioner` function is the one that links a vector with the inverse diagonal to the actual `DiagonalMatrix<VectorType>` variable, including the case where we access the matrix entries through `el(i,i)`.